## Intermediate Data Science

#### University of Redlands - DATA 201
#### Prof: Joanna Bieri [joanna_bieri@redlands.edu](mailto:joanna_bieri@redlands.edu)
#### [Class Website: data201.joannabieri.com](https://joannabieri.com/data201_intermediate.html)

## Computer Set Up

In [122]:
!python --version

Python 3.13.5


In [123]:
!conda --version

conda 25.7.0


In [124]:
!git --version

git version 2.51.0.windows.1


**Clone the Repo for our Class**

<a href="https://github.com/Redlands-DATA201/FALL25" target="_blank">Redlands-DATA201/FALL2025</a>

In [125]:
# # Depending on your setup you might need to install modules
# !conda install -y numpy
# !conda install -y pandas
# !conda install -y matplotlib
# !conda install -y plotly

In [126]:
# Some basic package imports
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.io as pio
pio.renderers.defaule = 'colab'

# Review of DATA 101

For our review we are going to jump in and do some Exploratory Data Analysis (EDA) on a data set that we have seen before. This time you are just given a .csv file and your goal is to analyze it - answering these questions.

**I expect this to feel impossible at first!!!** but if you all work together, share code, and look things up as needed, I know you can do it!

#### This was a Homework Assignment in Data 101 - but see how much you can figure out on your own!

* Answer the following questions using reproducible Python code.
    - What does it mean to be reproducible? This means someone else at any time in the future can run and understand your code almost like reading a blog post.
* For each question, state your answer in a sentence, e.g. "In this sample, the first three common names of purchasers are ...".
* Note that the answers to all questions are within the context of this particular sample of sales, i.e. you shouldn't make inferences about the population of all Lego sales based on this sample.

0.  Read in the .csv file using Pandas and display it

1.  Describe what you see in the data set (variables, observations, etc)

2.  What are the three most common first names of purchasers?

3.  What are the three most common themes of Lego sets purchased?

4.  Among the most common theme of Lego sets purchased, what is the most common subtheme?

5.  Create data frames for each of the ages in the following categories: "18 and under", "19 - 25", "26 - 35", "36 - 50", "51 and over". HINT - use masks

6.  Which age group has purchased the highest number of Lego sets.

7.  Which age group has spent the most money on Legos?

8.  Which Lego theme has made the most money for Lego?

9.  Which area code has spent the most money on Legos? In the US the area code is the first 3 digits of a phone number. Then using a for loop calculate the average money spent per customer for each area code.

10.  Come up with a question you want to answer using these data, and write it down. Then, create a data visualization that answers the question, and explain how your visualization answers the question.

**This exersize comes from: https://datasciencebox.org/course-materials/hw-instructions/hw-05/hw-05-legos**

#### 0. Read in the .csv file using Pandas and display it.

In [127]:
file_name = 'data/lego-sales.csv'

In [128]:
DF = pd.read_csv(file_name)
DF

first_name  last_name  age  phone_number  set_id   number  \
0     Kimberly  Beckstead   24  216-555-2549   24701    76062   
1         Neel     Garvin   35  819-555-3189   25626    70595   
2         Neel     Garvin   35  819-555-3189   24665    21031   
3      Chelsea   Bouchard   41           NaN   24695    31048   
4      Chelsea   Bouchard   41           NaN   25626    70595   
..         ...        ...  ...           ...     ...      ...   
615     Talise   Nieukirk   16  801-555-2343   24902    41556   
616    Spencer     Morgan   28  784-555-3455   26041    41580   
617    Spencer     Morgan   28  784-555-3455   26060  5005051   
618     Amelia    Hageman   40  336-555-1950   24702    76063   
619     Amelia    Hageman   40  336-555-1950   24720    10830   

                      theme              subtheme  year  \
0    DC Comics Super Heroes         Mighty Micros  2018   
1                   Ninjago  Rise of the Villains  2018   
2              Architecture                   NaN  2018   
3                   Creator                   NaN  2018   
4                   Ninjago  Rise of the Villains  2018   
..                      ...                   ...   ...   
615                  Mixels              Series 7  2018   
616                  Mixels              Series 9  2018   
617                    Gear         Digital Media  2018   
618  DC Comics Super Heroes         Mighty Micros  2018   
619                   Duplo                   NaN  2018   

                                       name  pieces  us_price  \
0                            Robin vs. Bane    77.0      9.99   
1                      Ultra Stealth Raider  1093.0    119.99   
2                              Burj Khalifa   333.0     39.99   
3                            Lakeside Lodge   368.0     29.99   
4                      Ultra Stealth Raider  1093.0    119.99   
..                                      ...     ...       ...   
615                                  Tiketz    62.0      4.99   
616                                    Myke    63.0      4.99   
617  Friends of Heartlake City Girlz 4 Life     NaN     19.99   
618              The Flash vs. Captain Cold    88.0      9.99   
619                           Minnie's Café    27.0     19.99   

                                             image_url  quantity  
0    http://images.brickset.com/sets/images/76062-1...         1  
1    http://images.brickset.com/sets/images/70595-1...         1  
2    http://images.brickset.com/sets/images/21031-1...         1  
3    http://images.brickset.com/sets/images/31048-1...         1  
4    http://images.brickset.com/sets/images/70595-1...         1  
..                                                 ...       ...  
615  http://images.brickset.com/sets/images/41556-1...         2  
616                                                NaN         2  
617                                                NaN         1  
618  http://images.brickset.com/sets/images/76063-1...         2  
619  http://images.brickset.com/sets/images/10830-1...         4  

[620 rows x 14 columns]

#### 1. Describe what you see in the data set (variables, observations, etc).

In [129]:
DF.shape

(620, 14)

In [130]:
DF.columns

Index(['first_name', 'last_name', 'age', 'phone_number', 'set_id', 'number',
       'theme', 'subtheme', 'year', 'name', 'pieces', 'us_price', 'image_url',
       'quantity'],
      dtype='object')

In [131]:
DF.dtypes

first_name       object
last_name        object
age               int64
phone_number     object
set_id            int64
number           object
theme            object
subtheme         object
year              int64
name             object
pieces          float64
us_price        float64
image_url        object
quantity          int64
dtype: object

In [132]:
DF.describe()

age        set_id    year       pieces    us_price    quantity
count  620.000000    620.000000   620.0   551.000000  620.000000  620.000000
mean    34.356452  25124.982258  2018.0   254.206897   29.041613    1.437097
std     11.276537    506.760720     0.0   357.738804   34.630623    0.712849
min     16.000000  24548.000000  2018.0    13.000000    3.990000    1.000000
25%     25.000000  24724.750000  2018.0    70.000000    9.990000    1.000000
50%     33.000000  24804.500000  2018.0   114.000000   19.990000    1.000000
75%     41.000000  25640.250000  2018.0   313.000000   29.990000    2.000000
max     68.000000  26060.000000  2018.0  4634.000000  349.990000    5.000000

There are 620 observations and 14 variables. The variables are 'first_name' (object), 'last_name' (object), 'age' (int), 'phone_number' (object, xxx-xxx-xxxx), 'set_id' (int), 'number' (object), 'theme' (object), 'subtheme' (object), 'year' (int), 'name' (object), 'pieces' (float), 'us_price' (float), 'image_url' (object), 'quantity' (int).

#### 2. What are the three most common first names of purchasers?

In [133]:
my_cols = ['first_name']
DF[my_cols].value_counts()

first_name
Jackson       13
Jacob         11
Joseph        11
Michael       10
Audrey         8
              ..
Sonny          1
Talise         1
Stephanie      1
Trevon         1
Van            1
Name: count, Length: 211, dtype: int64

In this sample, the three most common first names of purchasers are Jackson (13 observations), Jacob (11 observations), and Joseph (11 observations).

#### 3. What are the three most common themes of Lego sets purchased?

In [134]:
my_cols = ['theme']
DF[my_cols].value_counts()

theme                  
Star Wars                  75
Nexo Knights               64
Gear                       55
Mixels                     55
City                       45
Friends                    42
Ninjago                    38
Duplo                      35
Bionicle                   34
Creator                    25
DC Comics Super Heroes     22
Elves                      22
Marvel Super Heroes        19
Dimensions                 18
Disney Princess            15
The Angry Birds Movie      11
Architecture               10
Technic                    10
Minecraft                   9
Advanced Models             4
Collectable Minifigures     3
Seasonal                    3
Ghostbusters                3
Ideas                       2
Classic                     1
Name: count, dtype: int64

In this sample, the three most common themes of Lego sets purchased are Star Wars (75 observations), Nexo Knights (64 observations), and Gear (55 observations)/Mixels (55 observations).

#### 4. Among the most common theme of Lego sets purchased, what is the most common subtheme?

In [135]:
my_cols = ['theme','subtheme']
mask = DF['theme']=='Star Wars'
DF[mask][my_cols].value_counts()

theme      subtheme                 
Star Wars  The Force Awakens            15
           Buildable Figures            11
           Episode V                    10
           MicroFighters                10
           Battlefront                   7
           Original Content              7
           Episode III                   6
           Seasonal                      3
           Rebels                        3
           Episode IV                    2
           Ultimate Collector Series     1
Name: count, dtype: int64

In this sample, the most common subtheme of the most common theme of Lego sets purchased (Star Wars) is Episode III.

#### 5. Create data frames for each of the ages in the following categories: "18 and under", "19 - 25", "26 - 35", "36 - 50", "51 and over". HINT - use masks

In [136]:
my_cols = 'age'
mask = DF[my_cols]<=18
DF_18 = DF[mask].copy()
DF_18

first_name          last_name  age  phone_number  set_id   number  \
44    Michelle          Uguccioni   17           NaN   24756    70310   
45    Michelle          Uguccioni   17           NaN   24896    31046   
70       Lucas  Jimenez-Dominguez   18  712-555-0459   24691    31043   
71       Lucas  Jimenez-Dominguez   18  712-555-0459   26034    41573   
72       Lucas  Jimenez-Dominguez   18  712-555-0459   24704    41140   
73       Lucas  Jimenez-Dominguez   18  712-555-0459   24903    41558   
74       Lucas  Jimenez-Dominguez   18  712-555-0459   26033    41572   
187   Chayanne           Williams   17  869-555-6681   24678    60109   
188   Chayanne           Williams   17  869-555-6681   24697    76044   
223    Michael               Cruz   17  361-555-8212   24723    41172   
224    Michael               Cruz   17  361-555-8212   24618    71238   
225    Michael               Cruz   17  361-555-8212   24920    70605   
284     Steven         Washington   16  336-555-1433   25910    30347   
285     Steven         Washington   16  336-555-1433   25893    30349   
286     Serena         Bartkowiak   16  305-555-2377   24806    71311   
287     Serena         Bartkowiak   16  305-555-2377   24778    75113   
288     Serena         Bartkowiak   16  305-555-2377   24987    40201   
439      Aaron             Bruner   18  914-555-5003   24679    60111   
440      Aaron             Bruner   18  914-555-5003   24956    71300   
441      Aaron             Bruner   18  914-555-5003   24796    75137   
443    Michael             Donner   18  432-555-3069   24660    71241   
444    Michael             Donner   18  432-555-3069   24709    10803   
445    Michael             Donner   18  432-555-3069   26059  5005060   
476      Dylan            Mcgrath   18  229-555-1970   24748    76050   
477      Dylan            Mcgrath   18  229-555-1970   25645    71314   
611    Natalia           Vonduyke   16  859-555-3818   25879   853534   
612    Natalia           Vonduyke   16  859-555-3818   24714    10814   
613    Natalia           Vonduyke   16  859-555-3818   24765    70332   
614    Natalia           Vonduyke   16  859-555-3818   25922    75149   
615     Talise           Nieukirk   16  801-555-2343   24902    41556   

                      theme                            subtheme  year  \
44             Nexo Knights                                 NaN  2018   
45                  Creator                                 NaN  2018   
70                  Creator                                 NaN  2018   
71                   Mixels                            Series 9  2018   
72          Disney Princess                         Palace Pets  2018   
73                   Mixels                            Series 7  2018   
74                   Mixels                            Series 9  2018   
187                    City                                Fire  2018   
188  DC Comics Super Heroes  Batman v Superman: Dawn of Justice  2018   
223                   Elves                                 NaN  2018   
224              Dimensions                            Fun Pack  2018   
225                 Ninjago                            Skybound  2018   
284                    City                                Fire  2018   
285                    City                                 NaN  2018   
286                Bionicle                       Miscellaneous  2018   
287               Star Wars                   Buildable Figures  2018   
288                Seasonal                     Valentine's Day  2018   
439                    City                                Fire  2018   
440                Bionicle                          companions  2018   
441               Star Wars                           Episode V  2018   
443              Dimensions                            Fun Pack  2018   
444                   Duplo                                 NaN  2018   
445                    Gear                     Video Games/3

In [137]:
my_cols = 'age'
mask = (DF[my_cols]>=19) & (DF[my_cols]<=25)
DF_19 = DF[mask].copy()
DF_19

first_name    last_name  age  phone_number  set_id number  \
0     Kimberly    Beckstead   24  216-555-2549   24701  76062   
6      Bryanna        Welsh   19           NaN   24797  75138   
7      Bryanna        Welsh   19           NaN   24701  76062   
10       Chase  Fortenberry   19  205-555-3704   24707  10801   
11       Chase  Fortenberry   19  205-555-3704   24713  10809   
..         ...          ...  ...           ...     ...    ...   
591      Paige          Ice   21           NaN   26038  41577   
592      Aiden       Ganley   19  786-555-5067   25639  70324   
593      Aiden       Ganley   19  786-555-5067   24756  70310   
608    Carolyn       Quarry   23  567-555-7649   25627  70596   
609    Carolyn       Quarry   23  567-555-7649   24806  71311   

                      theme              subtheme  year  \
0    DC Comics Super Heroes         Mighty Micros  2018   
6                 Star Wars             Episode V  2018   
7    DC Comics Super Heroes         Mighty Micros  2018   
10                    Duplo                   NaN  2018   
11                    Duplo                   NaN  2018   
..                      ...                   ...   ...   
591                  Mixels              Series 9  2018   
592            Nexo Knights                   NaN  2018   
593            Nexo Knights                   NaN  2018   
608                 Ninjago  Rise of the Villains  2018   
609                Bionicle         Miscellaneous  2018   

                             name  pieces  us_price  \
0                  Robin vs. Bane    77.0      9.99   
6                     Hoth Attack   233.0     24.99   
7                  Robin vs. Bane    77.0      9.99   
10                   Baby Animals    13.0      9.99   
11                  Police Patrol    15.0     14.99   
..                            ...     ...       ...   
591                         Mysto    64.0      4.99   
592          Merlok's Library 2.0   288.0     24.99   
593       Knighton Battle Blaster    76.0      9.99   
608          Samurai X Cave Chaos  1253.0    119.99   
609  Kopaka and Melum - Unity set   171.0     24.99   

                                             image_url  quantity  
0    http://images.brickset.com/sets/images/76062-1...         1  
6    http://images.brickset.com/sets/images/75138-1...         1  
7    http://images.brickset.com/sets/images/76062-1...         3  
10   http://images.brickset.com/sets/images/10801-1...         1  
11   http://images.brickset.com/sets/images/10809-1...         1  
..                                                 ...       ...  
591                                                NaN         1  
592  http://images.brickset.com/sets/images/70324-1...         1  
593  http://images.brickset.com/sets/images/70310-1...         2  
608                                                NaN         2  
609  http://images.brickset.com/sets/images/71311-1...         2  

[129 rows x 14 columns]

In [138]:
my_cols = 'age'
mask = (DF[my_cols]>=26) & (DF[my_cols]<=35)
DF_26 = DF[mask].copy()
DF_26

first_name  last_name  age  phone_number  set_id   number         theme  \
1         Neel     Garvin   35  819-555-3189   25626    70595       Ninjago   
2         Neel     Garvin   35  819-555-3189   24665    21031  Architecture   
27       Jacob  Nzabanita   31  339-555-2572   24732    41117       Friends   
28       Jacob  Nzabanita   31  339-555-2572   24793    75133     Star Wars   
29       Jacob  Nzabanita   31  339-555-2572   24723    41172         Elves   
..         ...        ...  ...           ...     ...      ...           ...   
606   Benjamin       Park   33           NaN   24957    71309      Bionicle   
607      Miles       Hill   34           NaN   24758    70312  Nexo Knights   
610   Jennifer    Reinert   29  402-555-0467   24660    71241    Dimensions   
616    Spencer     Morgan   28  784-555-3455   26041    41580        Mixels   
617    Spencer     Morgan   28  784-555-3455   26060  5005051          Gear   

                 subtheme  year                                    name  \
1    Rise of the Villains  2018                    Ultra Stealth Raider   
2                     NaN  2018                            Burj Khalifa   
27               Pop Star  2018                      Pop Star TV Studio   
28            Battlefront  2018              Rebel Alliance Battle Pack   
29                    NaN  2018              The Water Dragon Adventure   
..                    ...   ...                                     ...   
606                   Toa  2018                  Onua - Uniter of Earth   
607                   NaN  2018                     Lance's Mecha Horse   
610              Fun Pack  2018                        Fun Pack: Slimer   
616              Series 9  2018                                    Myke   
617         Digital Media  2018  Friends of Heartlake City Girlz 4 Life   

     pieces  us_price                                          image_url  \
1    1093.0    119.99  http://images.brickset.com/sets/images/70595-1...   
2     333.0     39.99  http://images.brickset.com/sets/images/21031-1...   
27    194.0     19.99  http://images.brickset.com/sets/images/41117-1...   
28    101.0     12.99  http://images.brickset.com/sets/images/75133-1...   
29    212.0     19.99  http://images.brickset.com/sets/images/41172-1...   
..      ...       ...                                                ...   
606   143.0     19.99  http://images.brickset.com/sets/images/71309-1...   
607   237.0     19.99  http://images.brickset.com/sets/images/70312-1...   
610    33.0     14.99  http://images.brickset.com/sets/images/71241-1...   
616    63.0      4.99                                                NaN   
617     NaN     19.99                                                NaN   

     quantity  
1           1  
2           1  
27          1  
28          2  
29          1  
..        ...  
606         2  
607         2  
610         3  
616         2  
617         1  

[183 rows x 14 columns]

In [139]:
my_cols = 'age'
mask = (DF[my_cols]>=36) & (DF[my_cols]<=50)
DF_36 = DF[mask].copy()
DF_36

first_name       last_name  age  phone_number  set_id number  \
3      Chelsea        Bouchard   41           NaN   24695  31048   
4      Chelsea        Bouchard   41           NaN   25626  70595   
5      Chelsea        Bouchard   41           NaN   24721  10831   
8        Caleb  Garcia-Wideman   37  907-555-9236   24730  41115   
9        Caleb  Garcia-Wideman   37  907-555-9236   25611  21127   
..         ...             ...  ...           ...     ...    ...   
598     Payton          Milsap   42           NaN   24702  76063   
599     Payton          Milsap   42           NaN   24678  60109   
602  Stephanie        Harrison   42           NaN   24736  41121   
618     Amelia         Hageman   40  336-555-1950   24702  76063   
619     Amelia         Hageman   40  336-555-1950   24720  10830   

                      theme              subtheme  year  \
3                   Creator                   NaN  2018   
4                   Ninjago  Rise of the Villains  2018   
5                     Duplo                   NaN  2018   
8                   Friends                   NaN  2018   
9                 Minecraft         Minifig-scale  2018   
..                      ...                   ...   ...   
598  DC Comics Super Heroes         Mighty Micros  2018   
599                    City                  Fire  2018   
602                 Friends        Adventure Camp  2018   
618  DC Comics Super Heroes         Mighty Micros  2018   
619                   Duplo                   NaN  2018   

                           name  pieces  us_price  \
3                Lakeside Lodge   368.0     29.99   
4          Ultra Stealth Raider  1093.0    119.99   
5          My First Caterpillar    19.0      9.99   
8      Emma's Creative Workshop   108.0      9.99   
9                  The Fortress     NaN    109.99   
..                          ...     ...       ...   
598  The Flash vs. Captain Cold    88.0      9.99   
599                   Fire Boat   412.0     79.99   
602      Adventure Camp Rafting   320.0     29.99   
618  The Flash vs. Captain Cold    88.0      9.99   
619               Minnie's Café    27.0     19.99   

                                             image_url  quantity  
3    http://images.brickset.com/sets/images/31048-1...         1  
4    http://images.brickset.com/sets/images/70595-1...         1  
5    http://images.brickset.com/sets/images/10831-1...         1  
8    http://images.brickset.com/sets/images/41115-1...         1  
9    http://images.brickset.com/sets/images/21127-1...         2  
..                                                 ...       ...  
598  http://images.brickset.com/sets/images/76063-1...         3  
599  http://images.brickset.com/sets/images/60109-1...         2  
602  http://images.brickset.com/sets/images/41121-1...         1  
618  http://images.brickset.com/sets/images/76063-1...         2  
619  http://images.brickset.com/sets/images/10830-1...         4  

[216 rows x 14 columns]

In [140]:
my_cols = 'age'
mask = DF[my_cols]>=51
DF_51 = DF[mask].copy()
DF_51

first_name      last_name  age  phone_number  set_id number       theme  \
31      Hannah  Drews Stunkel   55  339-555-6320   25628  70590     Ninjago   
32      Hannah  Drews Stunkel   55  339-555-6320   25624  70593     Ninjago   
33      Hannah  Drews Stunkel   55  339-555-6320   24734  41119     Friends   
38       Riley            Ott   51  517-555-2093   24785  75126   Star Wars   
39       Riley            Ott   51  517-555-2093   24772  70604     Ninjago   
..         ...            ...  ...           ...     ...    ...         ...   
552      Angel          Payne   62           NaN   25605  75146   Star Wars   
553      Angel          Payne   62           NaN   24793  75133   Star Wars   
594     Dustin      Vanvuuren   51  812-555-1009   24709  10803       Duplo   
595     Dustin      Vanvuuren   51  812-555-1009   24807  71235  Dimensions   
596     Dustin      Vanvuuren   51  812-555-1009   25920  75151   Star Wars   

          subtheme  year                        name  pieces  us_price  \
31             NaN  2018     Airjitzu Battle Grounds   666.0     59.99   
32        Skybound  2018        The Green NRG Dragon   567.0     49.99   
33             NaN  2018      Heartlake Cupcake Cafe   439.0     39.99   
38   MicroFighters  2018     First Order Snowspeeder    91.0      9.99   
39        Skybound  2018          Tiger Widow Island   450.0     49.99   
..             ...   ...                         ...     ...       ...   
552       Seasonal  2018   Star Wars Advent Calendar   282.0     39.99   
553    Battlefront  2018  Rebel Alliance Battle Pack   101.0     12.99   
594            NaN  2018                      Arctic    40.0     29.99   
595     Level Pack  2018   Level Pack: Midway Arcade    96.0     29.99   
596    Episode III  2018            Clone Turbo Tank   903.0    109.99   

                                             image_url  quantity  
31                                                 NaN         1  
32   http://images.brickset.com/sets/images/70593-1...         3  
33   http://images.brickset.com/sets/images/41119-1...         1  
38   http://images.brickset.com/sets/images/75126-1...         2  
39   http://images.brickset.com/sets/images/70604-1...         1  
..                                                 ...       ...  
552  http://images.brickset.com/sets/images/75146-1...         4  
553  http://images.brickset.com/sets/images/75133-1...         1  
594  http://images.brickset.com/sets/images/10803-1...         2  
595  http://images.brickset.com/sets/images/71235-1...         1  
596  http://images.brickset.com/sets/images/75151-1...         1  

[62 rows x 14 columns]

#### 6. Which age group has purchased the highest number of Lego sets.

In [141]:
DF['age'].min()

16

In [142]:
DF['age'].max()

68

In [143]:
DF['age_group'] = pd.cut(DF['age'], bins=[15,18,25,35,50,68], labels=['16-18','19-25','26-35','36-50','51-68'])
DF

first_name  last_name  age  phone_number  set_id   number  \
0     Kimberly  Beckstead   24  216-555-2549   24701    76062   
1         Neel     Garvin   35  819-555-3189   25626    70595   
2         Neel     Garvin   35  819-555-3189   24665    21031   
3      Chelsea   Bouchard   41           NaN   24695    31048   
4      Chelsea   Bouchard   41           NaN   25626    70595   
..         ...        ...  ...           ...     ...      ...   
615     Talise   Nieukirk   16  801-555-2343   24902    41556   
616    Spencer     Morgan   28  784-555-3455   26041    41580   
617    Spencer     Morgan   28  784-555-3455   26060  5005051   
618     Amelia    Hageman   40  336-555-1950   24702    76063   
619     Amelia    Hageman   40  336-555-1950   24720    10830   

                      theme              subtheme  year  \
0    DC Comics Super Heroes         Mighty Micros  2018   
1                   Ninjago  Rise of the Villains  2018   
2              Architecture                   NaN  2018   
3                   Creator                   NaN  2018   
4                   Ninjago  Rise of the Villains  2018   
..                      ...                   ...   ...   
615                  Mixels              Series 7  2018   
616                  Mixels              Series 9  2018   
617                    Gear         Digital Media  2018   
618  DC Comics Super Heroes         Mighty Micros  2018   
619                   Duplo                   NaN  2018   

                                       name  pieces  us_price  \
0                            Robin vs. Bane    77.0      9.99   
1                      Ultra Stealth Raider  1093.0    119.99   
2                              Burj Khalifa   333.0     39.99   
3                            Lakeside Lodge   368.0     29.99   
4                      Ultra Stealth Raider  1093.0    119.99   
..                                      ...     ...       ...   
615                                  Tiketz    62.0      4.99   
616                                    Myke    63.0      4.99   
617  Friends of Heartlake City Girlz 4 Life     NaN     19.99   
618              The Flash vs. Captain Cold    88.0      9.99   
619                           Minnie's Café    27.0     19.99   

                                             image_url  quantity age_group  
0    http://images.brickset.com/sets/images/76062-1...         1     19-25  
1    http://images.brickset.com/sets/images/70595-1...         1     26-35  
2    http://images.brickset.com/sets/images/21031-1...         1     26-35  
3    http://images.brickset.com/sets/images/31048-1...         1     36-50  
4    http://images.brickset.com/sets/images/70595-1...         1     36-50  
..                                                 ...       ...       ...  
615  http://images.brickset.com/sets/images/41556-1...         2     16-18  
616                                                NaN         2     26-35  
617                                                NaN         1     26-35  
618  http://images.brickset.com/sets/images/76063-1...         2     36-50  
619  http://images.brickset.com/sets/images/10830-1...         4     36-50  

[620 rows x 15 columns]

In [144]:
my_cols = ['age_group','quantity']
DF[my_cols].groupby('age_group').sum().sort_values('quantity',ascending=False)

C:\Users\josee\AppData\Local\Temp\ipykernel_2976\817112597.py:2: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



quantity
age_group          
36-50           313
26-35           267
19-25           174
51-68            92
16-18            45

In this sample, the 36-50 age group has purchased the highest number of Lego sets.

#### 7. Which age group has spent the most money on Legos?

In [145]:
DF['us_revenue'] = DF['us_price']*DF['quantity']
DF

first_name  last_name  age  phone_number  set_id   number  \
0     Kimberly  Beckstead   24  216-555-2549   24701    76062   
1         Neel     Garvin   35  819-555-3189   25626    70595   
2         Neel     Garvin   35  819-555-3189   24665    21031   
3      Chelsea   Bouchard   41           NaN   24695    31048   
4      Chelsea   Bouchard   41           NaN   25626    70595   
..         ...        ...  ...           ...     ...      ...   
615     Talise   Nieukirk   16  801-555-2343   24902    41556   
616    Spencer     Morgan   28  784-555-3455   26041    41580   
617    Spencer     Morgan   28  784-555-3455   26060  5005051   
618     Amelia    Hageman   40  336-555-1950   24702    76063   
619     Amelia    Hageman   40  336-555-1950   24720    10830   

                      theme              subtheme  year  \
0    DC Comics Super Heroes         Mighty Micros  2018   
1                   Ninjago  Rise of the Villains  2018   
2              Architecture                   NaN  2018   
3                   Creator                   NaN  2018   
4                   Ninjago  Rise of the Villains  2018   
..                      ...                   ...   ...   
615                  Mixels              Series 7  2018   
616                  Mixels              Series 9  2018   
617                    Gear         Digital Media  2018   
618  DC Comics Super Heroes         Mighty Micros  2018   
619                   Duplo                   NaN  2018   

                                       name  pieces  us_price  \
0                            Robin vs. Bane    77.0      9.99   
1                      Ultra Stealth Raider  1093.0    119.99   
2                              Burj Khalifa   333.0     39.99   
3                            Lakeside Lodge   368.0     29.99   
4                      Ultra Stealth Raider  1093.0    119.99   
..                                      ...     ...       ...   
615                                  Tiketz    62.0      4.99   
616                                    Myke    63.0      4.99   
617  Friends of Heartlake City Girlz 4 Life     NaN     19.99   
618              The Flash vs. Captain Cold    88.0      9.99   
619                           Minnie's Café    27.0     19.99   

                                             image_url  quantity age_group  \
0    http://images.brickset.com/sets/images/76062-1...         1     19-25   
1    http://images.brickset.com/sets/images/70595-1...         1     26-35   
2    http://images.brickset.com/sets/images/21031-1...         1     26-35   
3    http://images.brickset.com/sets/images/31048-1...         1     36-50   
4    http://images.brickset.com/sets/images/70595-1...         1     36-50   
..                                                 ...       ...       ...   
615  http://images.brickset.com/sets/images/41556-1...         2     16-18   
616                                                NaN         2     26-35   
617                                                NaN         1     26-35   
618  http://images.brickset.com/sets/images/76063-1...         2     36-50   
619  http://images.brickset.com/sets/images/10830-1...         4     36-50   

     us_revenue  
0          9.99  
1        119.99  
2         39.99  
3         29.99  
4        119.99  
..          ...  
615        9.98  
616        9.98  
617       19.99  
618       19.98  
619       79.96  

[620 rows x 16 columns]

In [146]:
my_cols = ['age_group','us_revenue']
DF[my_cols].groupby('age_group').sum().sort_values('us_revenue',ascending=False)

C:\Users\josee\AppData\Local\Temp\ipykernel_2976\3391705059.py:2: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



us_revenue
age_group            
36-50         9532.87
26-35         7576.33
19-25         4939.26
51-68         2475.08
16-18          948.55

In this sample, the 36-50 age group has spent the most money on Legos ($9,532.87).

#### 8. Which Lego theme has made the most money for Lego?

In [147]:
my_cols = ['theme','us_revenue']
DF[my_cols].groupby('theme').sum().sort_values('us_revenue',ascending=False)

us_revenue
theme                              
Star Wars                   4447.83
Ninjago                     2279.41
City                        2211.35
Nexo Knights                2209.07
Minecraft                   1549.90
Gear                        1533.21
Friends                     1279.44
Duplo                       1219.50
Elves                       1119.72
Ghostbusters                 879.95
Advanced Models              849.95
Marvel Super Heroes          839.70
Bionicle                     779.54
DC Comics Super Heroes       653.65
Architecture                 589.85
Technic                      572.88
Creator                      549.65
Dimensions                   484.76
The Angry Birds Movie        478.87
Mixels                       414.17
Disney Princess              237.80
Ideas                        209.97
Seasonal                      39.96
Classic                       29.99
Collectable Minifigures       11.97

In this sample, the Star Wars theme has made the most money for Lego ($4,447.83).

#### 9. Which area code has spent the most money on Legos? In the US the area code is the first 3 digits of a phone number. Then using a for loop calculate the average money spent per customer for each area code.

In [148]:
print(DF['phone_number'].isna().sum())

92


In [149]:
DF['area_code'] = DF['phone_number'].apply(lambda x: str(x).split('-')[0])
DF

first_name  last_name  age  phone_number  set_id   number  \
0     Kimberly  Beckstead   24  216-555-2549   24701    76062   
1         Neel     Garvin   35  819-555-3189   25626    70595   
2         Neel     Garvin   35  819-555-3189   24665    21031   
3      Chelsea   Bouchard   41           NaN   24695    31048   
4      Chelsea   Bouchard   41           NaN   25626    70595   
..         ...        ...  ...           ...     ...      ...   
615     Talise   Nieukirk   16  801-555-2343   24902    41556   
616    Spencer     Morgan   28  784-555-3455   26041    41580   
617    Spencer     Morgan   28  784-555-3455   26060  5005051   
618     Amelia    Hageman   40  336-555-1950   24702    76063   
619     Amelia    Hageman   40  336-555-1950   24720    10830   

                      theme              subtheme  year  \
0    DC Comics Super Heroes         Mighty Micros  2018   
1                   Ninjago  Rise of the Villains  2018   
2              Architecture                   NaN  2018   
3                   Creator                   NaN  2018   
4                   Ninjago  Rise of the Villains  2018   
..                      ...                   ...   ...   
615                  Mixels              Series 7  2018   
616                  Mixels              Series 9  2018   
617                    Gear         Digital Media  2018   
618  DC Comics Super Heroes         Mighty Micros  2018   
619                   Duplo                   NaN  2018   

                                       name  pieces  us_price  \
0                            Robin vs. Bane    77.0      9.99   
1                      Ultra Stealth Raider  1093.0    119.99   
2                              Burj Khalifa   333.0     39.99   
3                            Lakeside Lodge   368.0     29.99   
4                      Ultra Stealth Raider  1093.0    119.99   
..                                      ...     ...       ...   
615                                  Tiketz    62.0      4.99   
616                                    Myke    63.0      4.99   
617  Friends of Heartlake City Girlz 4 Life     NaN     19.99   
618              The Flash vs. Captain Cold    88.0      9.99   
619                           Minnie's Café    27.0     19.99   

                                             image_url  quantity age_group  \
0    http://images.brickset.com/sets/images/76062-1...         1     19-25   
1    http://images.brickset.com/sets/images/70595-1...         1     26-35   
2    http://images.brickset.com/sets/images/21031-1...         1     26-35   
3    http://images.brickset.com/sets/images/31048-1...         1     36-50   
4    http://images.brickset.com/sets/images/70595-1...         1     36-50   
..                                                 ...       ...       ...   
615  http://images.brickset.com/sets/images/41556-1...         2     16-18   
616                                                NaN         2     26-35   
617                                                NaN         1     26-35   
618  http://images.brickset.com/sets/images/76063-1...         2     36-50   
619  http://images.brickset.com/sets/images/10830-1...         4     36-50   

     us_revenue area_code  
0          9.99       216  
1        119.99       819  
2         39.99       819  
3         29.99       nan  
4        119.99       nan  
..          ...       ...  
615        9.98       801  
616        9.98       784  
617       19.99       784  
618       19.98       336  
619       79.96       336  

[620 rows x 17 columns]

In [150]:
DF['area_code'].value_counts()

area_code
nan    92
281     9
712     9
567     9
441     9
       ..
509     1
419     1
787     1
520     1
402     1
Name: count, Length: 157, dtype: int64

In [151]:
print(DF[my_cols].isna().sum())

theme         0
us_revenue    0
dtype: int64


In [152]:
my_cols = ['us_revenue','area_code']
DF_area = DF[my_cols].dropna().copy()
DF_area

us_revenue area_code
0          9.99       216
1        119.99       819
2         39.99       819
3         29.99       nan
4        119.99       nan
..          ...       ...
615        9.98       801
616        9.98       784
617       19.99       784
618       19.98       336
619       79.96       336

[620 rows x 2 columns]

#### 10. Come up with a question you want to answer using these data, and write it down. Then, create a data visualization that answers the question, and explain how your visualization answers the question.